services/bundle_analysis/new_notify/__init__.py

```diff
@@ -0,0 +1,164 @@
+)
+from services.bundle_analysis.new_notify.helpers import (
+    get_notification_types_configured,
+)
+from services.bundle_analysis.new_notify.messages import MessageStrategyInterface
+from services.bundle_analysis.new_notify.messages.comment import (
+    BundleAnalysisCommentMarkdownStrategy,
+)
+from services.bundle_analysis.new_notify.types import NotificationType
+
+log = logging.getLogger(__name__)
+
+
+class NotificationFullContext(NamedTuple):
+    notification_context: BaseBundleAnalysisNotificationContext
+    message_strategy: MessageStrategyInterface
+
+
+class BundleAnalysisNotifyReturn(NamedTuple):
+    notifications_configured: tuple[NotificationType]
+    notifications_successful: tuple[NotificationType]
+
+
+class BundleAnalysisNotifyService:
+    def __init__(
+        self,
+        commit: Commit,
+        current_yaml: UserYaml,
+        gh_app_installation_name: str | None = None,
+    ):
+        self.commit = commit
+        self.current_yaml = current_yaml
+        self.gh_app_installation_name = (
+            gh_app_installation_name or GITHUB_APP_INSTALLATION_DEFAULT_NAME
+        )
+
+    @property
+    def owner(self) -> Owner:
+        return self.commit.repository.owner
+
+    def build_base_context(self) -> BaseBundleAnalysisNotificationContext | None:
+        try:
+            return (
+                NotificationContextBuilder()
+                .initialize(
+                    self.commit, self.current_yaml, self.gh_app_installation_name
+                )
+                .build_context()
+                .get_result()
+            )
+        except NotificationContextBuildError as exp:
+            log.error(
+                "Failed to build NotificationContext",
+                extra=dict(
+                    notification_type="base_context", failed_step=exp.failed_step
+                ),
+            )
+        return None
+
+    def create_context_for_notification(
+        self,
+        base_context: BaseBundleAnalysisNotificationContext,
+        notification_type: NotificationType,
+    ) -> NotificationFullContext | None:
+        """Builds the NotificationContext for the given notification_type
+        If the NotificationContext failed to build we can't send this notification.
+
+        Each NotificationType is paired with a ContextBuilder and MessageStrategyInterface.
+        The MessageStrategy is later used to build and send the message based on the NotificationContext
+        """
+        builders_lookup: dict[
+            NotificationType,
+            tuple[NotificationContextBuilder, MessageStrategyInterface],
+        ] = {
+            NotificationType.PR_COMMENT: (
+                BundleAnalysisPRCommentContextBuilder,
+                BundleAnalysisCommentMarkdownStrategy,
+            )
+        }
+        builder_class, message_strategy_class = builders_lookup.get(
+            notification_type, (None, None)
+        )
+
+        if builder_class is None:
```

Comment:
Is this supposed to be extended in the future?

If so: can you move that dict to the global scope, as it is fully static.
If not: this could be a simple `if notification_type is not NotificationType.PR_COMMENT:`, along with inlining the `message_strategy_class`.

Also, I’m not quite sure about this, but I don’t think `mypy` can reason about `message_strategy_class` based on `builder_class`, so it will likely complain that it can be `None` down below when calling it.
So it might be better to check for a non-None `.get()` result first before unpacking the tuple.

Reply:
yes, it's supposed to be extended in the future. I could use a `match` statement on `NotificationType` but I guess I prefer the dictionary approach